File: ainpc-core-plugin/src/main/java/ro/ainpc/engine/QuestDirectorRequest.java
package ro.ainpc.engine;

import ro.ainpc.progression.ProgressionDefinition;
import ro.ainpc.story.StoryContextSnapshot;

import java.util.ArrayList;
import java.util.List;

public record QuestDirectorRequest(
    StoryContextSnapshot storyContext,
    List<ProgressionDefinition> definitions,
    String preferredMechanicId,
    boolean questSeedAllowed,
    List<String> blockingReasons
) {
    public QuestDirectorRequest {
        storyContext = storyContext != null ? storyContext : StoryContextSnapshot.empty();
        definitions = List.copyOf(definitions != null ? definitions.stream()
            .filter(definition -> definition != null)
            .toList() : List.of());
        preferredMechanicId = valueOrEmpty(preferredMechanicId);
        blockingReasons = sanitizeStrings(blockingReasons);
    }

    public static QuestDirectorRequest forStoryContext(StoryContextSnapshot storyContext,
                                                       List<ProgressionDefinition> definitions) {
        return new QuestDirectorRequest(storyContext, definitions, "", false, List.of());
    }

    private static List<String> sanitizeStrings(List<String> values) {
        if (values == null || values.isEmpty()) {
            return List.of();
        }

        List<String> sanitized = new ArrayList<>();
        for (String value : values) {
            String safeValue = valueOrEmpty(value);
            if (!safeValue.isBlank()) {
                sanitized.add(safeValue);
            }
        }
        return List.copyOf(sanitized);
    }

    private static String valueOrEmpty(String value) {
        return value == null ? "" : value.trim();
    }
}

File: ainpc-core-plugin/src/main/kotlin/ro/ainpc/engine/QuestDirectorRequest.kt
package ro.ainpc.engine

import ro.ainpc.progression.ProgressionDefinition
import ro.ainpc.story.StoryContextSnapshot

class QuestDirectorRequest(
    storyContext: StoryContextSnapshot?,
    definitions: List<ProgressionDefinition>?,
    preferredMechanicId: String?,
    private val questSeedAllowedValue: Boolean,
    blockingReasons: List<String>?
) {
    private val storyContextValue: StoryContextSnapshot = storyContext ?: StoryContextSnapshot.empty()
    private val definitionsValue: List<ProgressionDefinition> =
        (definitions ?: emptyList()).filterNotNull().toList()
    private val preferredMechanicIdValue: String = valueOrEmpty(preferredMechanicId)
    private val blockingReasonsValue: List<String> = sanitizeStrings(blockingReasons)

    fun storyContext(): StoryContextSnapshot = storyContextValue
    fun definitions(): List<ProgressionDefinition> = definitionsValue
    fun preferredMechanicId(): String = preferredMechanicIdValue
    fun questSeedAllowed(): Boolean = questSeedAllowedValue
    fun blockingReasons(): List<String> = blockingReasonsValue

    companion object {
        @JvmStatic
        fun forStoryContext(
            storyContext: StoryContextSnapshot?,
            definitions: List<ProgressionDefinition>?
        ): QuestDirectorRequest = QuestDirectorRequest(storyContext, definitions, "", false, emptyList())

        private fun sanitizeStrings(values: List<String>?): List<String> {
            if (values.isNullOrEmpty()) {
                return emptyList()
            }
            val sanitized = ArrayList<String>()
            for (value in values) {
                val safeValue = valueOrEmpty(value)
                if (safeValue.isNotBlank()) {
                    sanitized.add(safeValue)
                }
            }
            return sanitized.toList()
        }

        private fun valueOrEmpty(value: String?): String = value?.trim().orEmpty()
    }
}
